Reloading DNS-over-QUIC crashes: ServePacket dereferences nil quicListener

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via security@coredns.io
-->

**What happened**:
Reloading a CoreDNS instance that serves DNS-over-QUIC causes a panic:

```
CoreDNS-1.13.1
linux/arm64, go1.25.2, 1db4568
[INFO] Reloading
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x67367c]
```
This is because tls is reinitiallized with an empty key. When putting the wrapper around it, the actual error occours:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x6a4544]
```
```
$ echo 0x6a4544| go tool addr2line ./coredns
github.com/coredns/coredns/core/dnsserver.(*ServerQUIC).ServePacket()
core/dnsserver/server_quic.go:<line>
github.com/quic-go/quic-go@v0.55.0/server.go:143
```
`ServePacket` dereferences a nil `s.quicListener` because the new server instance never rebuilt it before the reused UDP socket was handed over.

**What you expected to happen**:
Reload should reuse the existing QUIC socket without crashing.

**How to reproduce it (as minimally and precisely as possible)**:
1. Start CoreDNS with the Corefile below (QUIC + `reload`).
2. Modify the configmap, add a plugin for example
3. The process panics in `ServePacket`.

**Anything else we need to know?**:
On reload, Caddy skips a new `ListenPacket()` call and invokes `ServePacket()` directly with the old `net.PacketConn`. `ServePacket` assumes `s.quicListener` was initialised already, so the dereference crashes. Lazily wrapping the provided `PacketConn` when `s.quicListener` is nil fixes it; I have a patch ready.

**Environment**:

- the version of CoreDNS: 1.13.1
- Corefile:
```yaml
    quic://.:784 {
        tls /etc/coredns/ssl/tls.crt /etc/coredns/ssl/tls.key {
            client_auth nocert
        }
        quic
        reload
    }
```
- logs, if applicable: see panic trace above.
- OS (e.g: `cat /etc/os-release`): its the default coredns/coredns container

--- a/core/dnsserver/server_quic.go
+++ b/core/dnsserver/server_quic.go
@@ -103,6 +103,14 @@ func NewServerQUIC(addr string, group []*Config) (*ServerQUIC, error) {
 // ServePacket implements caddy.UDPServer interface.
 func (s *ServerQUIC) ServePacket(p net.PacketConn) error {
 	s.m.Lock()
+	if s.quicListener == nil {
+		listener, err := quic.Listen(p, s.tlsConfig, s.quicConfig)
+		if err != nil {
+			s.m.Unlock()
+			return err
+		}
+		s.quicListener = listener
+	}
 	s.listenAddr = s.quicListener.Addr()
 	s.m.Unlock()
 

--- a/test/quic_test.go
+++ b/test/quic_test.go
@@ -23,6 +23,12 @@ var quicCorefile = `quic://.:0 {
 		whoami
 	}`
 
+var quicReloadCorefile = `quic://.:0 {
+		tls ../plugin/tls/test_cert.pem ../plugin/tls/test_key.pem ../plugin/tls/test_ca.pem
+		whoami
+		reload 2s
+	}`
+
 // Corefile with custom stream limits
 var quicLimitCorefile = `quic://.:0 {
 		tls ../plugin/tls/test_cert.pem ../plugin/tls/test_key.pem ../plugin/tls/test_ca.pem
@@ -87,6 +93,29 @@ func TestQUIC(t *testing.T) {
 	if len(d.Extra) != 2 {
 		t.Errorf("Expected 2 RRs in additional section, but got %d", len(d.Extra))
 	}
+}
+
+func TestQUICReloadDoesNotPanic(t *testing.T) {
+	inst, udp, _, err := CoreDNSServerAndPorts(quicCorefile)
+	if err != nil {
+		t.Fatalf("Could not get CoreDNS serving instance: %s", err)
+	}
+	t.Cleanup(func() { inst.Stop() })
+
+	assertQUICQuerySucceeds(t, udp)
+
+	restart, err := inst.Restart(NewInput(quicReloadCorefile))
+	if err != nil {
+		t.Fatalf("Failed to restart CoreDNS: %s", err)
+	}
+	t.Cleanup(func() { restart.Stop() })
+
+	udpReload, _ := CoreDNSServerPorts(restart, 0)
+	if udpReload == "" {
+		t.Fatal("Failed to determine QUIC listener address after reload")
+	}
+
+	assertQUICQuerySucceeds(t, udpReload)
 }
 
 func TestQUICProtocolError(t *testing.T) {
@@ -351,4 +380,51 @@ func createInvalidDOQMsg() []byte {
 	m.SetQuestion("whoami.example.org.", dns.TypeA)
 	msg, _ := m.Pack()
 	return msg
+}
+
+func assertQUICQuerySucceeds(t *testing.T, address string) {
+	t.Helper()
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	conn, err := quic.DialAddr(ctx, convertAddress(address), generateTLSConfig(), nil)
+	if err != nil {
+		t.Fatalf("Expected no error but got: %s", err)
+	}
+	defer func() { _ = conn.CloseWithError(0, "") }()
+
+	stream, err := conn.OpenStreamSync(ctx)
+	if err != nil {
+		t.Fatalf("Expected no error but got: %s", err)
+	}
+	defer func() { _ = stream.Close() }()
+
+	msg := createTestMsg()
+	if _, err = stream.Write(msg); err != nil {
+		t.Fatalf("Expected no error but got: %s", err)
+	}
+
+	sizeBuf := make([]byte, 2)
+	if _, err = io.ReadFull(stream, sizeBuf); err != nil {
+		t.Fatalf("Expected no error but got: %s", err)
+	}
+
+	size := binary.BigEndian.Uint16(sizeBuf)
+	buf := make([]byte, size)
+	if _, err = io.ReadFull(stream, buf); err != nil {
+		t.Fatalf("Expected no error but got: %s", err)
+	}
+
+	resp := new(dns.Msg)
+	if err = resp.Unpack(buf); err != nil {
+		t.Fatalf("Expected no error but got: %s", err)
+	}
+
+	if resp.Rcode != dns.RcodeSuccess {
+		t.Fatalf("Expected success but got %d", resp.Rcode)
+	}
+
+	if len(resp.Extra) != 2 {
+		t.Fatalf("Expected 2 RRs in additional section, but got %d", len(resp.Extra))
+	}
 }

